compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -728,13 +728,12 @@ class Typer extends Namer
       case _: WildcardType => untpd.TypeTree()
       case _ => untpd.TypeTree(tp)
     }
-    pt.stripTypeVar match {
-      case _ if defn.isNonDepFunctionType(pt) =>
+    pt.dealias.followSyntheticOpaque match {
```

Comment:
This seems fishy to me, outside of the companion of the opaque type, we shouldn't be doing this. And inside the companion, it should not be needed (since we can just dealias). The real issue I think is that `isNonDepFunctionType` is overly strict in what it considers a function type, such that an intersection of a function type and something else is not consider a function type. If we relax this, then I don't think the `followSyntheticOpaque` is needed. (This is closely related to https://github.com/lampepfl/dotty/issues/5473 where the issue is that the definition of implicit function type is too restrictive).

Reply:
Dealiasing does not work inside the companion object, since for an opaque type 

    opaque type T = A

the binding of `T.T` has the form

    type T >: A | outer.T <: A & outer.T

That's why `followSyntheticOpaque` is needed; it uncovers the alias `A` when inside the companion object.

Reply:
But If I look at the base types of `T` in the companion object, I should see `Function1[A, Boolean]`, because base types follow upper bounds, so if `isNonDepFunctionType` used `isRef` or something like that, we wouldn't need the special case.

Reply:
It would have to be `derivesFrom` but that would break other things. The code here is quite finely balanced out, and I don't want to upset it since this might take a long time and the result is probably the same.

Reply:
I'm mostly mentioning this because https://github.com/lampepfl/dotty/issues/5473 could be solved the same way (using `derivesFrom`) and if we don't do that then more special cases are going to break (e.g. an implicit function type behind an opaque type)

Reply:
I propose to treat #5473 separately. I feel that's a much bigger can of worms than it looks at first.